### PR TITLE
First implementation

### PR DESCRIPTION
**Pull request by jonmagic**
@bkeepers @spraints this is the implementation for the [README](https://github.com/jonmagic/arca/blob/502cecdb57eb1f9253215a1fcafcc3a69f909d00/README.md) I showed you guys earlier today.

I haven't done any class or method documentation yet but I wanted to get a first version out so I could try it on some real projects and get a feel for how it works. I'd love feedback on the code if you have time to review at some point.
- [x] TomDoc methods
- [x] Add tests for:
  - [x] https://github.com/jonmagic/arca/commit/9798e1f8ced1b444569c6deef52603eef8301086
  - [x] https://github.com/jonmagic/arca/commit/8c3b8f347d7e7868dbe2660f8c2b89e54818a0f9
  - [x] https://github.com/jonmagic/arca/commit/3a00466c07044d5c949200a5763803b811c5dee3
  - [x] https://github.com/jonmagic/arca/commit/2f61776323cf39fd42a6c51c1647d7c15371af3a
  - [x] https://github.com/jonmagic/arca/commit/cdc1d9d6367689efcf57406a6177bad5e9124eaf
- [x] Get code review and iterate
- [x] Merge this PR
- [x] Cut v1.0.0 and publish to rubygems.org

/cc @kevinsawicki

### CODE REVIEW

**Comment by jonmagic** [comment]
@jnunemaker would you be interested in helping me with this PR by giving it some code review :eyes:

**Comment by bkeepers** [comment]
Nice! This looks really good. Nothing sticks out to me in the code.

File: lib/arca/collector.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                :callback_symbol                => callback_symbol,
+                :callback_file_path             => callback_line_matches[1],
+                :callback_line_number           => callback_line_matches[2].to_i,
+                :target_symbol                  => target_symbol,
+                :conditional_symbol             => conditional_symbol,
+                :conditional_target_symbol      => conditional_target_symbol
+              }
+
+              # Bind the callback method to self and call it with args.
+              callback_method.bind(self).call(*args)
```
Comment (spraints) [inline lib/arca/collector.rb]: This looks like it's not called right. Should this be outside of the `args.each` loop, and have `options` added back onto the args array?

**Comment by spraints** [comment]
This is really neat. :zap:

One thing that might make arca better is if it could live outside of the app. Rather than having it loaded into the running app all the time and using `rails console` to generate the report, what if there was an executable that you ran to generate the report? Something like "run `arca app/models/ticket.rb` to generate the report". This is more of a suggestion for arca as a whole, not this PR.

File: README.md [hunk 2]
```diff
@@ -10,20 +10,39 @@ The Arca library has two main components, the collector and the reporter. Includ
...
+```
+gem 'arca'
+```
+
+Add an initializer to require the library and configure it (`config/initializers/arca.rb` for example). There's no magic here and Arca doesn't assume your root project path or the path to your `ActiveRecord` models so you have to specify those paths yourself in the initializer.
+
+```
+require "arca"
+
+Arca.root_path = Rails.root
```
Comment (jnunemaker) [inline README.md:24]: I'd be tempted to add arca-rails gem that does this for you and includes Arca::Collector in ActiveRecord::Base.

File: lib/arca/callback_analysis.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    # callback including target methods, file paths, line numbers, booleans
+    # representing whether targets are in the same file they are called from,
+    # and finally the number of lines between callers and the target methods
+    # they call.
+    #
+    # model         - Arca::Model instance.
+    # callback_data - Hash with callback data collected by Arca::Collector.
+    def initialize(model:, callback_data:)
+      @model                          = model
+      @callback_symbol                = callback_data[:callback_symbol]
```
Comment (jnunemaker) [inline lib/arca/callback_analysis.rb]: Are these required?

File: lib/arca/model.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    def analyzed_callbacks
+      @analyzed_callbacks ||= CALLBACKS.inject({}) do |result, callback_symbol|
+        Array(callbacks[callback_symbol]).each do |callback_data|
+          result[callback_symbol] ||= []
+          callback_analysis = CallbackAnalysis.new({
+            :model         => self,
+            :callback_data => callback_data
+          })
+
+          result[callback_symbol] << callback_analysis unless callback_analysis.target_file_path =~ /gems\/activerecord/
```
Comment (jnunemaker) [inline lib/arca/model.rb]: Minor, but this is hard to read as it is long and wraps. I'd perhaps move the unless to the line before and indent the logic.

Also, minor but seems like you could have callback_analysis.target_file_path_active_record? or something to wrap this up.

File: lib/arca/report.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        :number_of_callbacks => number_of_callbacks,
+        :lines_between       => lines_between_callbacks,
+        :externals           => externals,
+        :conditionals        => conditionals,
+        :permutations        => permutations
+      }
+    end
+
+    # Public: String class name of model.
+    def model_class
```
Comment (jnunemaker) [inline lib/arca/report.rb]: I'd probably go with model_name as it is the String name not the class instance.

File: lib/arca/report.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    end
+
+    # Public: String file path of model.
+    def model_file_path
+      model.file_path
+    end
+
+    # Public: Integer representing the number of callbacks used in this model.
+    def number_of_callbacks
+      model.analyzed_callbacks_array.size
```
Comment (jnunemaker) [inline lib/arca/report.rb]: Seems like this could be wrapped up in model.analyzed_callbacks_count.

File: lib/arca/report.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      sorted_line_numbers = line_numbers.sort {|a,b| b <=> a }
+      sorted_line_numbers.each_with_index do |line_number, index|
+        lines_between += line_number - (sorted_line_numbers[index + 1] || 0)
+      end
+      lines_between
+    end
+
+    # Public: Integer representing the number of callbacks called for this class
+    # from files other than the one where the class is defined.
+    def externals
```
Comment (jnunemaker) [inline lib/arca/report.rb]: This name confused me when I just saw the report hash. I assumed it was a callback that made a call to an external service. I don't have a better name, but thought it was worth mentioning. All the names I can think of are really long: non_source_file_callbacks, external_file_callbacks, etc.

File: lib/arca/report.rb [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      model.analyzed_callbacks_array.select do |analysis|
+        analysis.external_callback? || analysis.external_target? || analysis.external_conditional_target?
```
Comment (jnunemaker) [inline lib/arca/report.rb]: Seems like analysis could have an external method to wrap this up.

File: lib/arca/report.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    def conditionals
+      number_of_unique_conditionals(model.analyzed_callbacks_array)
+    end
+
+    # Public: Integer representing the possible number of permutations stemming
+    # from conditionals for an instance of the model being reported during the
+    # lifecycle of the object.
+    def permutations
+      permutations = model.analyzed_callbacks.inject([]) do |results, (key, analyzed_callbacks)|
+        results << 2**number_of_unique_conditionals(analyzed_callbacks)
```
Comment (jnunemaker) [inline lib/arca/report.rb]: 2 can't breathe yo. Perhaps some spaces would help. 2 *\* number...

File: lib/arca/report.rb [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+    # Internal: Integer representing the number of unique conditions for an
+    # Array of CallbackAnalysis objects.
+    #
+    # analyzed_callbacks - Array of CallbackAnalysis objects.
+    #
+    # Returns an Integer.
+    def number_of_unique_conditionals(analyzed_callbacks)
+      analyzed_callbacks.
+        select {|analysis| analysis.conditional_symbol }.
```
Comment (jnunemaker) [inline lib/arca/report.rb:74]: Rando, but I hate this style. Feel free to ignore, but If you are going to have a space before the trailing, you should have a space after the leading bracket.

File: lib/arca.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+require_relative "arca/model"
+require_relative "arca/report"
+require_relative "arca/callback_analysis"
+
+module Arca
+
+  # Public: Reader method for accessing the Arca::Model for analysis and
+  # reporting.
+  def self.[](klass)
+    Arca::Model.new(klass)
```
Comment (jnunemaker) [inline lib/arca.rb:20]: May want to guard against string/symbol class or require kind_of?(Class) or something to fail early if someone passes in the wrong type.

File: arca.gemspec [hunk 12]
```diff
@@ -0,0 +1,23 @@
+Gem::Specification.new do |spec|
+  spec.name          = "arca"
+  spec.version       = "1.0.0"
+  spec.date          = "2015-07-11"
+  spec.summary       = "ActiveRecord callback analyzer"
+  spec.description   = "Arca is a callback analyzer for ActiveRecord ideally suited for digging yourself out of callback hell"
+  spec.authors       = ["Jonathan Hoyt"]
+  spec.email         = "jonmagic@gmail.com"
+  spec.require_paths = ["lib"]
+  spec.files         = [
```
Comment (jnunemaker) [inline arca.gemspec]: I like the bundler style of using git:

```
spec.files = `git ls-files -z`.split("\x0")
spec.executables = spec.files.grep(%r{^bin/}) { |f| File.basename(f) }
spec.test_files = spec.files.grep(%r{^(test|spec|features)/})
```

**Comment by jnunemaker** [comment]
Mostly minor stuff. Looks good. I like @spraints idea of running outside rails. I could also see a plugin with a slick web ui where you just drop it in and go to /arca and you can click on all models and see graphs and all kinds of fun things.

File: lib/arca/collector.rb (hunk 1 above)
Comment (jonmagic) [inline lib/arca/collector.rb]: Fixed in c1ca7aded862ade3adba0509162398fecf03721b

File: lib/arca/callback_analysis.rb (hunk 3 above)
Comment (jonmagic) [inline lib/arca/callback_analysis.rb]: Yes they are needed. Model is used to calculate several values later in the class and callback_symbol is included in the callback analysis hash. I did remove the extra spaces though in 0d3c4d245db4eeb29f32b8d65377edc9a49712c3

File: lib/arca/model.rb (hunk 4 above)
Comment (jonmagic) [inline lib/arca/model.rb]: Fixed in 2e6573fd06749cb59a6a75c5961a0036f63f7d37

File: lib/arca/report.rb (hunk 5 above)
Comment (jonmagic) [inline lib/arca/report.rb]: Fixed in 11e445527884fb869c3a1873b0c9e76305f4daa4

File: lib/arca/report.rb (hunk 10 above)
Comment (jonmagic) [inline lib/arca/report.rb:74]: I'm partial to the style I used here after trying all of the others over the years. This way still reads easiest for me.

File: lib/arca/report.rb (hunk 7 above)
Comment (jonmagic) [inline lib/arca/report.rb]: In e44e70e5d0646b5d8974cd69874ab4dd1250af02 I went with included_callbacks_count and renamed some other things to try and improve clarity. Thoughts?

File: lib/arca/report.rb (hunk 6 above)
Comment (jonmagic) [inline lib/arca/report.rb]: I did a bunch of refactoring in 8735b1d6042979eca3579fc56a29a0dbecafd485

File: lib/arca/report.rb (hunk 7 above)
Comment (jonmagic) [inline lib/arca/report.rb]: In 8735b1d6042979eca3579fc56a29a0dbecafd485 I went a step further and broke the included_callbacks_count into three separate counts.
- included_callbacks_count - callbacks included from another file
- external_targets_count - callback targets that are defined in a different file from the callback
- external_conditionals_count - callback conditional targets that are defined in a different file from the callback

File: lib/arca.rb (hunk 11 above)
Comment (jonmagic) [inline lib/arca.rb:20]: Updated in 2b6db5da319e5ab5f1be723d386c4834a341580e and also in fcf797cb1c1db264398a502c66009129baf0af51 I now raise early if the class passed in hasn't had any data collected for it.

File: arca.gemspec (hunk 12 above)
Comment (jonmagic) [inline arca.gemspec]: Updated in 78401d87b8a8aa28572d8da893e85154de5cd2c9

**Comment by jonmagic** [comment]
I'm not going to do the rails gem, executable, or web report for the first version. For the gem's 1.0.0 trail run I'm going to use it in a larger app and write some automated lint tests to prevent existing code from getting worse.

File: lib/arca/callback_analysis.rb (hunk 3 above)
Comment (jnunemaker) [inline lib/arca/callback_analysis.rb]: I should have explained. If they are required then you should probably use callback_data.fetch(:callback_symbol). That is what I was going for.

Comment (jonmagic) [inline lib/arca/callback_analysis.rb]: Ahh, I should have caught that :) Thanks man, updated in a4a291a7434ede9a6722f057d06bf1127c737366

**Pull request merged**